tests/test_mssql_rawquery_block.py

```diff
@@ -0,0 +1,136 @@
+            'PORT={};'
+            'SERVER={};'
+            'DATABASE={};'
+            'UID={};'
+            'MARS_Connection={};'
+            'PWD={}'.format(
+                self._driver,
+                self._port,
+                self._host,
+                self._db,
+                self._uid,
+                'yes',
+                self._pw))
+        self.assertEqual(mock_cnxn.cursor.call_count, 1)
+        mock_cursor.execute.assert_called_once_with('SELECT * FROM ? WHERE foo = ? AND pi > ?', ['a_table', 'bar', 3])
+        self.assertEqual(mock_cursor.close.call_count, 1)
+        self.assertEqual(mock_cnxn.close.call_count, 1)
+
+    @patch(MSSQLBase.__module__ + '.pyodbc')
+    def test_null_results(self, mock_odbc):
+        mock_cnxn = mock_odbc.connect.return_value = MagicMock()
+        mock_cursor = mock_cnxn.cursor.return_value = MagicMock()
+        mock_cursor.execute.return_value = mock_cursor
+        mock_cursor.fetchall.return_value = []
+        mock_cursor.columns.return_value = [
+            MagicMock(column_name='foo'),
+            MagicMock(column_name='pi')]
+        blk = MSSQLRawQuery()
+        self.configure_block(blk, self.config)
+        blk.start()
+        blk.process_signals([Signal({'headers': self.config['headers'], 'query': self.config['query'], 'parameters': self.config['parameters']})])
+        blk.stop()
+        self.assert_num_signals_notified(1)
+        self.assertDictEqual(
+          self.last_notified['no_results'][0].to_dict(),
+          {'results': 'null', 'headers': { 'foo': 'bar' }, 'query': 'SELECT * FROM ? WHERE foo = ? AND pi > ?', 'parameters': ['a_table', 'bar', 3]}
+        )
+        mock_odbc.connect.assert_called_once_with(
+            'DRIVER={};'
+            'PORT={};'
+            'SERVER={};'
+            'DATABASE={};'
+            'UID={};'
+            'MARS_Connection={};'
+            'PWD={}'.format(
+                self._driver,
+                self._port,
+                self._host,
+                self._db,
+                self._uid,
+                'yes',
+                self._pw))
+        self.assertEqual(mock_cnxn.cursor.call_count, 1)
+        mock_cursor.execute.assert_called_once_with('SELECT * FROM ? WHERE foo = ? AND pi > ?', ['a_table', 'bar', 3])
+        self.assertEqual(mock_cursor.close.call_count, 1)
+        self.assertEqual(mock_cnxn.close.call_count, 1)
+
+    @patch(MSSQLBase.__module__ + '.pyodbc')
+    def test_exception_handling(self, mock_odbc):
+        mock_cnxn = mock_odbc.connect.return_value = MagicMock()
+        mock_cursor = mock_cnxn.cursor.side_effect = Exception()
+        blk = MSSQLRawQuery()
+        self.configure_block(blk, self.config)
+        blk.start()
+        self.assertEqual(mock_odbc.connect.call_count, 1)
+        with self.assertRaises(Exception):
+            blk.process_signals([Signal({'query': 'SELECT * FROM ? WHERE foo = ? AND pi > ?', 'parameters': ['a_table', 'bar', 3]})])

```

Comment:
This test shows exactly why we shouldn't be processing queries off of signal attributes. A signal containing a SQL query that gets executed is about the scariest thing I can imagine seeing. 😱

Reply:
this has been fixed

Reply:
tangentally-related to this, I don't like that [this signal](https://github.com/nio-blocks/mssql/blob/master/mssql_update_block.py#L58) generated in the update block uses the key of `"Rows updated"`. It just makes it annoying to get out a value, whether it's python or javascript.